src/engine/index.js

```
@@ -77,35 +86,73 @@ function id() {
   return _.uniqueId('switch-to-uuid-later') // TODO we need to use UUID
 }
 
+const indexes = new WeakMap()
+
+function ensureIndex(array, property) {
+  if (indexes.has(array) && indexes.get(array).property != null) {
+    return
+  } else {
+    const index = {}
+    for (let elem of array) {
+      index[elem[property]] = elem
+    }
+    if (!indexes.has(array)) {
+      indexes.set(array, {})
+    }
+    indexes.get(array)[property] = index
+  }
+}
+
+function getBy(array, property, value) {
+  ensureIndex(array, property)
+  const index = indexes.get(array)[property]
+  return index[value]
+}
+
+function getById(array, value) {
+  return getBy(array, 'id', value)
+}
+
 function cloneData(data) {
-  let res = Object.create(null)
 
-  // Clone Deep won't copy array properties.
-  for (let tabId in data) {
-    const index = _.cloneDeep(data[tabId].id)
-    res[tabId] = _.values(index)
-    res[tabId].id = index
+  function customizer(value) {
+    if (errorData.has(value)) {
+      return {...errorData.get(value), [errorSentinel]: true}
+    } else {
+      return undefined
+    }
   }
 
-  return res
+  return _.cloneDeepWith(data, customizer)
 }
 
 function setCells({schema, data}, cells) {
   data = cloneData(data)
 
+  // build temporary index, so we can process input cells effectively
+  for (let tabId in data) {
+    // for computed tables, tabData may be error object
+    if (schema[tabId].fn != null) continue
+    const tabData = data[tabId]
+    tabData.id = {}
+    for (let row of tabData) {
+      tabData.id[row.id] = row
+    }
+  }
+
   for (let [tabId, rowId, colId, value] of cells) {
     if (schema[tabId].fn || schema[tabId].columns[colId].fn) continue
     _.setWith(data, [tabId, 'id', rowId, colId], value, (x) => x || Object.create(null))
   }
 
   for (let tabId in data) {
+    if (schema[tabId].fn != null) continue
```

This can't happen. If `schema[tabId].fn != null`, it is error. So, if you want to check for that, feel free, but `throw` instead of `continue`.